src/python/grpcio/grpc/experimental/aio/_channel.py

```diff
@@ -88,22 +101,77 @@ def __call__(self,
 
         serialized_request = _common.serialize(request,
                                                self._request_serializer)
-        timeout = self._timeout_to_deadline(timeout)
         aio_cancel_status = cygrpc.AioCancelStatus()
-        aio_call = asyncio.ensure_future(
-            self._channel.unary_unary(self._method, serialized_request, timeout,
-                                      aio_cancel_status),
-            loop=self._loop)
+
+        if self._interceptors:
+            client_call_details = _ClientCallDetails(
+                self._method, timeout, metadata, credentials, wait_for_ready,
+                compression)
+            aio_call = asyncio.ensure_future(
+                self._wrap_call_in_interceptors(
+                    client_call_details, serialized_request, aio_cancel_status),
+                loop=self._loop)
+        else:
+            aio_call = asyncio.ensure_future(
+                self._call(self._method, serialized_request, timeout,
+                           aio_cancel_status),
+                loop=self._loop)
+
         return Call(aio_call, self._response_deserializer, aio_cancel_status)
 
+    async def _wrap_call_in_interceptors(
+            self, client_call_details: ClientCallDetails, request: bytes,
+            aio_cancel_status: cygrpc.AioCancelStatus):
+        """Run the RPC call wraped in interceptors"""
+
+        async def _run_interceptor(
+                interceptors: Iterable[UnaryUnaryClientInterceptor],
+                client_call_details: ClientCallDetails, request: bytes):
+            try:
+                interceptor = next(interceptors)
+            except StopIteration:
+                interceptor = None
+
+            if interceptor:
+                continuation = functools.partial(_run_interceptor, interceptors)
+                return await interceptor.intercept_unary_unary(
+                    continuation, client_call_details, request)
+            else:
+                return await self._call(client_call_details.method, request,
+                                        client_call_details.timeout,
+                                        aio_cancel_status)
```

Comment:
I think the overall design is fine, but can we make `_run_interceptor` a normal function that returns `Call` object? A potential way of doing that:

```Python
def _run_interceptor(
        interceptors: Iterable[UnaryUnaryClientInterceptor],
        client_call_details: ClientCallDetails, request: bytes):
    try:
        interceptor = next(interceptors)
    except StopIteration:
        interceptor = None

    if interceptor:
        continuation = functools.partial(_run_interceptor, interceptors)
        aio_call = asyncio.ensure_future(
            interceptor.intercept_unary_unary(
                continuation,
                client_call_details,
                request
            )
        )
    else:
        asyncio.ensure_future(
            self._call(client_call_details.method,
                       request,
                       client_call_details.timeout,
                       aio_cancel_status),
        )
    return Call(aio_call, self._response_deserializer, aio_cancel_status)
```

Reply:
Despite this would solve one of the main problems that I mentioned, having the interceptor not forced to make two `awaits` for resolving the `Call` since the continuation function is not a coroutine,  the main down-side effect of this would be having to create as many Asyncio tasks as many interceptors we have, which defenitly might impact in the performance.

I can measure what would be the impact in terms of QPS and see if there is any substantial degradation.

But I do not dislike this idea, because it would provide consistency in the API and allowing cancellation.

Thanks!

Reply:
I remember I performed a benchmark to measure the difference between directly `await` and create an `Task`. The cost should be similar since underlying they are all wrappers of the `call_soon` API. But as we know, the concrete benchmark result is always better than assumptions. We can make the decision based on the result.

```
Spent 16.08s for 1000000 runs of directly_await
16082.14 ns
Spent 15.65s for 1000000 runs of schedule_task
15653.08 ns
```

```Python
import asyncio
import timeit
import time

async def work():
    return "I am not working"

async def directly_await():
    await work()

def schedule_task():
    return asyncio.create_task(work())

async def main():
    number = 1000000
    start_time = time.time()
    await asyncio.gather(*(directly_await() for i in range(number)))
    elapsed_time = time.time() - start_time
    print("Spent %.2fs for %d runs of %s" % (elapsed_time, number, 'directly_await'))
    print("%.2f ns" % (elapsed_time / number * 1e9))
    
    start_time = time.time()
    await asyncio.gather(*(schedule_task() for i in range(number)))
    elapsed_time = time.time() - start_time
    print("Spent %.2fs for %d runs of %s" % (elapsed_time, number, 'schedule_task'))
    print("%.2f ns" % (elapsed_time / number * 1e9))


if __name__ == "__main__":
    asyncio.run(main())
```

Reply:
Umm, be careful the `gather` internally wraps coros into tasks [1], so you are comparing the same pattern implicitly, no direct call at all.

Ive changed a bit the script for comparing both patterns, see the following script and the results

```python
import asyncio
import timeit
import time

async def work():
    return "I am not working"

async def directly_await():
    await work()

async def schedule_task():
    return await asyncio.create_task(work())

async def main():
    number = 1000000
    start_time = time.time()
    for i in range(number):
        await directly_await()
    elapsed_time = time.time() - start_time
    print("Spent %.2fs for %d runs of %s" % (elapsed_time, number, 'directly_await'))
    print("%.2f ns" % (elapsed_time / number * 1e9))

    start_time = time.time()
    for i in range(number):
        await schedule_task()
    elapsed_time = time.time() - start_time
    print("Spent %.2fs for %d runs of %s" % (elapsed_time, number, 'schedule_task'))
    print("%.2f ns" % (elapsed_time / number * 1e9))


if __name__ == "__main__":
    asyncio.run(main())
```

The results are:

``` bash
$ python /tmp/coros_vs_tasks.py
Spent 0.33s for 1000000 runs of directly_await
332.50 ns
Spent 16.96s for 1000000 runs of schedule_task
16958.59 ns
```

Wrapping the coros in tasks means having a slow down of the code ~ x50 times. I guess that the `await`/`yield from` has a really small footprint since it can be processed by the CPython in one just one bytecode operation and some x86 instructions while creating a task and scheduling means a lot of Python code.


[1] https://github.com/python/cpython/blob/master/Lib/asyncio/tasks.py#L806

Reply:
If this is the case, then we might want to re-design the `grpc.aio.Call` object, since it already introduced a significant slowdown.

Reply:
This problem is going to bite us quickly, without access to the `Call` object, we will find the functionality of interceptors worse than normal call, e.g. no access to `cancel`, `initial_metadata`, `is_active`.

Reply:
```Python
async def create_future():
    future = loop.create_future()
    future.set_result(await work())
    return future
```
```
Spent 0.36s for 1000000 runs of directly_await
356.81 ns
Spent 0.92s for 1000000 runs of create_future
915.44 ns
```

A viable workaround can be using lower-level API to directly create future in the event loop. The performance suffers a bit, but provides better functionality. We can change the `Call` from wrapping a `Task` to a lower-level `future`.

Reply:
There is something that I'm missing out for sure, sorry for jeopardizing this thread with another question, but it's quite related.

**does the gRPC library provide a way for returning the initial metadata, payload and trailing metadata when they are available?** So, not having to wait until the whole RPC has finished [1].

This question IMO is fundamental, while in frameworks like `Aiohttp` the headers can be, for example, _awaited_ without having to _await_ for the full response I did not see this in the gRPC library, if it does exist my bad.

So considering a **No** for my question - which I might be totally wrong - my idea was that the current payload that is currently returned by the Cython call [1] would be eventually changed for returning more attributes, having something like this:

```python
if receive_status_on_client_operation.code() == StatusCode.ok:
    return (receive_message_operation.message(), receive_initial_metadata_operation.initial_metadata() ,....)
```

So the coroutine published by the Cython code would return not only the response but also the other metadata related to that call, this would mean that the interceptors would receive these other attributes too, most likely by converting this tuple to a kind of `ResponseDetails` object` here [2], so the interceptors would be able to do something like this.

```python
async def intercept_unary_unary(self, continuation,
                                        client_call_details, request):
    interceptors_executed.append(self)
    response_detail = await continuation(client_call_details, request)
    print(response_detail.initial_metadata())
    print(response_detail.trailing_metadata())
    return response_detail
```

So interceptors would have always the chance for accessing other attributes.

Am I wrong and exists an interface for gathering the metadata and response when they are available independently?

PD: In any case, I'm starting to think that the plumbing that is being done between the Python code and the Cython code for canceling an ongoing RPC is not the right one, having the feeling that the way for making it more robust and for the future has the chance for reading partial response data - like initial metadata - we would need to change a bit the plumbing, it would mean removing the current `AioCancelStatus` [3] class and returning the whole `AioCall` object and use it within the `Call` instance for canceling an RPC..

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi#L142
[2] https://github.com/grpc/grpc/pull/21246/files#diff-b44bc9a6efb8a6ba5b7e5037b6863b2cR148-R161
[3] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/aio/cancel_status.pyx.pxi#L17

Reply:
> does the gRPC library provide a way for returning the initial metadata, payload and trailing metadata when they are available?

Well, as you already familiar with the C-Core API, the operation you mentioning above are done by separated operations. E.g. [receiving initial metadata](https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/grpc_types.h#L567), [receiving final status including trailing metadata](https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/grpc_types.h#L577). If we are coding for non unary-unary cases, the operations usually are sent in multiple batches (see [code](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_channel.py#L737)). This usage is the intended usage of the `grpc_call_start_batch` API, which allows the wrapper writer to have fine grained control. One more implementation details about how the initial_metadata is set, see [code](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_channel.py#L132). Receiving individual payloads or trailing metadata have similar design.

From API perspective, the [`_Rendezvous`](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_channel.py#L351) class has implemented a blocking logic that calls to `initial_metadata`, `code`, etc. will block until the resource is available. In unary case, I agree, they are not particularly useful, but in streaming cases they are quite handy.

> Proposal of `response_detail`.

Personally, I'm a fan of simplicity. The root of the pain we have today (in this thread, and in gRFC) is the design of the related classes. I totally agree that if we introduce the new concept, the design will be much simpler, just like `requests` and `aiohttp`. For HTTP/1.1, the message and status arrived in the same time. But HTTP/2 is more or less a streaming protocol, so we need to give application a chance to perform business logic between the arrival of initial headers, payloads, and trailing headers.

In code, the unary-unary case may looks a bit complex.

```Python
async def intercept_unary_unary(self, continuation,
                                        client_call_details, request):
    interceptors_executed.append(self)
    call = continuation(client_call_details, request)
    print(await call.initial_metadata())
    print(await call.trailing_metadata())
    return await response_detail
```

But if we introduce streaming on either side, it will look more nature.

```Python
async def intercept_unary_stream(self, continuation,
                                        client_call_details, request):
    interceptors_executed.append(self)
    call = continuation(client_call_details, request)
    try:
        _CONFIGURATE_TRACING(await call.initial_metadata())
    except:
        call.cancel()
        raise
    else:
        async def response in call:
            yield response
```

> PD: In any case, I'm starting to think that the plumbing that is being done between the Python code and the Cython code for canceling an ongoing RPC is not the right one, having the feeling that the way for making it more robust and for the future has the chance for reading partial response data - like initial metadata - we would need to change a bit the plumbing, it would mean removing the current AioCancelStatus [3] class and returning the whole AioCall object and use it within the Call instance for canceling an RPC..

I have similar feeling, and experimenting alternatives in my streaming PR. In short, my current choice is using a future to allow the application communicate with the Cython layer, so the `cancel` can happen in any other coroutine (especially user's), and being handled quickly.

Reply:
Ok, now I got the point of this line [1] which I did not understand. So in the synchronous client are created as many tags - in our use case would be different callbacks - as many batch operations and it's up to the driver decide what operations are batched together - I guess that a batch is considered done when all operations that belong to a specific batch are finished.

Oks, if this is the case I do think also that the use case of returning everything in one shot is fundamentally broken considering that data could be already there.

My next question then would be, **what would need to be plumbed?** So considering that both - the caller and the interceptors - have to have access to different data - initial_metdata, , trailing_metadata, response, others? - that might be received in different moments how they and its respective methods would need to be chained? how is this being solved in other languages like Java, Go or C++? what happeen's with the current Python client?

The easy part is the code executed before making the call, as all code interceptors are being executed before making the final call we can say that the following would be what would happen in the real world

```
caller -- > interceptor 1 (work before the call is done) --> Interceptor N (work before the call is done) --> final call
```

True that because how is it designed right now, the caller would receive any early exception raised in the code executed before making the call once the caller would be trying to access the response or any of the caller attributes. Take as an example the following code where the caller receives an exception from an interceptor that implements a circuit breaker.

```python
call = Hi(....)
try:
   await call
except CicruitOpen:
    logging.warning("Circuit open for the stub Hi!!!!")
    raise
```

The `CicruitOpen` exception would be indeed raised in the code executed before making the call, but by design would be raised later.

The none easy part is what would happen with the code executed right after the interceptor. There are at least two options. 

Option 1: Every step needs to wait the finalization of the previous step. So for the following code:

```python
async def interceptor_2(request, clientDetails, continuation)
    call = await(interceptor)
    response = await call
    return call

async def interceptor_1(request, clientDetails, continuation)
    call = await(interceptor)
    await call.metadata()
    return call

async def caller(..)
    call = hi(...)
    await call.metadata()
```

Both the `caller` and the `interceptor_1` though only were accessing the metadata - that could be already there, both would need to wait till the response is already there because the `interceptor_2` blocked everything since this last interceptor was waiting for the final response.

Option2, only the `response` is being wired, so `metadata` can be accessed when it's available from any place. In that case, the previous code both - the `caller` and the `interceptor_1`- would have access to the `metadata`.  But one of the likely down-side effects by allowing this would be the chances of having data races in the client-side, for example:

```
async def interceptor_1(request, clientDetails, continuation)
    call = await(interceptor)
    metadata = await call.metadata()
    try:
        await do_something_async()
    except:
       call.cancel()
       raise
    return call

async def caller(..)
    call = hi(...)
    metadata = await call.metadata()  
```

Nothing would guarantee you that the `caller` wouldn't see the metadata for a call that would be eventually canceled by one of the interceptors. 

Thoughts? Having the feeling that simpler the better, so going for option 1 would need to be the way to go.

[1] https://github.com/grpc/grpc/blob/6950e15882f28e43685e948a7e5227bfcef398cd/src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi#L249

Reply:
@pfreixes I like the idea of option 2. It's much less rigid and allows certain interceptors to execute and terminate early (e.g. when initial metadata arrives). You bring up a valid point about data races. But it seems this could be fixed with asyncio locking primitives.

In your example above, we would take an `asyncio.Lock` in `call.metadata().__await__`. Then, in `interceptor_1` we would hold the lock between when we read the metadata and when we write to it (or cancel).

Order becomes more important in this model though, because if you actually want a particular interceptor to do something early, like when initial metadata arrives, it will need to be placed early in the interceptor pipeline. If an interceptor that awaits the whole RPC comes before it, then it will not execute early.

Reply:
Interceptors are executed in order and are implicitly chained when they are given during the channel instantiation, so I guess that by design order is predictable and there is no way for having a data race between interceptors.

So I guess that the only place where the Option 2 would be used will be between the caller and the interceptors, where the caller would have full freedom on reading the metadata once it is available instead of having to wait till all interceptors have finished.

Regarding the example of the lock just let me put an argument against this pattern, while the usage of a lock might have sense within the caller code having it between an interceptor and the caller would be weird to me, where most likely both pieces of code would have nothing in common, being maintained by different owners.

In any case, I would make the changes without adding any restriction and allowing the client/caller to read the metadata at any moment.